### PR TITLE
Uplift outlives components to `rustc_type_ir`

### PR DESCRIPTION
**Pull request by compiler-errors**
We need this to uplift `push_outlives_components`, since the elaborator uses `push_outlives_components` to elaborate type outlives obligations and I want to uplift elaboration.

This ends up reworking and inlining a fair portion of the `GenericArg::walk_shallow` function, whose only callsite was this one. I believe I got the logic correct, but may be worthwhile to look at it closely just in case. Unfortunately github was too dumb to understand that this is a rename + change -- I could also rework the git history to split the "copy the file over" part from the actual logical changes if that makes this easier to review.

r? lcnr

### CODE REVIEW

File: compiler/rustc_type_ir/src/outlives.rs [hunk 1]
```diff
@@ -0,0 +1,335 @@
+// The outlines relation `T: 'a` or `'a: 'b`. This code frequently
+// refers to rules defined in RFC 1214 (`OutlivesFooBar`), so see that
+// RFC for reference.
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/outlives.rs:3]:
```suggestion
//! The outlines relation `T: 'a` or `'a: 'b`. This code frequently
//! refers to rules defined in RFC 1214 (`OutlivesFooBar`), so see that
//! RFC for reference.
```

**Review by lcnr — COMMENTED** [review]
while I feel like this doesn't necessary have to happen in this PR, I really think this should use a `TypeVisitor` instead. Computing the outlives components for trait objects is fairly non-trivial.

r=me on the existing changes

**Comment by compiler-errors** [comment]
I could turn this into a TypeVisitor. The exceptions should be fairly obvious in that case.

**Comment by compiler-errors** [comment]
I'll do it in another PR tho since it's actually kinda gnarly because of this `visited` hashset we pass around

@bors r=lcnr

**Pull request merged**